### ISSUE TITLE
Training hang at start when running on multi-gpu

### ISSUE THREAD

**Issue by karan6181** [mosaicml/streaming#145]
## Environment

- OS: [Ubuntu 20.04]
- Hardware (GPU, or instance type): [A100, H100]

## To reproduce

When trying to run the model training on multi-gpu, I observed a hang. Running any multi-gpu model training produces a hang

Snapshot (curtailed output)

```
eval_dataset.resize_size: 232
model.loss_name: binary_cross_entropy
train_dataset.crop_size: 176

Run evaluation
******************************
Config:
blurpool/num_blurconv_layers: 6
blurpool/num_blurpool_layers: 1
num_gpus_per_node: 8
num_nodes: 1
rank_zero_seed: 17

******************************
<HANGED>
```

Supporting tracking JIRA issue: CO-1688

**Comment by karan6181** [comment]
Root cause:

- Without setting the seed inside the main script, all the ranks generates a unique [prefix_int](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataset.py#L205) which is not expected and this leads to [worker_barrier](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataset.py#L215) being hanged since it expects all the ranks to generate the same `prefix_int`.
- Setting the seed inside the main script ensures all the ranks generates the same prefix_int and there is no hang. Hence, if you are using [composer](https://github.com/mosaicml/composer/), then setting `composer.utils.reproducibility.seed_all(seed)` fixes the hang or if you are not using the composer, then simply doing  `np.random.seed(seed)` will fix the hang issue.

The above is the workaround solution for now until we fix the issue in the Streaming Dataset which will be very soon! So stay tuned for the update!

**Comment by karan6181** [comment]
PR which fixes this issue: https://github.com/mosaicml/streaming/pull/147
